crates/scsi/src/virtio.rs

```diff
@@ -0,0 +1,358 @@
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+//! Helpers for virtio and virtio-scsi.
+use std::{
+    cell::Cell,
+    cmp::{max, min},
+    convert::TryInto,
+    io,
+    io::{ErrorKind, Read, Write},
+    mem,
+    ops::Deref,
+    rc::Rc,
+};
+
+use log::error;
+
+use virtio_bindings::virtio_scsi::virtio_scsi_cmd_req;
+use virtio_queue::{Descriptor, DescriptorChain, DescriptorChainRwIter};
+use vm_memory::{Bytes, GuestAddress, GuestMemory};
+
+/// virtio-scsi has its own format for LUNs, documented in 5.6.6.1 of virtio
+/// v1.1. This represents a LUN parsed from that format.
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum VirtioScsiLun {
+    ReportLuns,
+    TargetLun(u8, u16),
+}
+
+const REPORT_LUNS: [u8; 8] = [0xc1, 0x01, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0];
+
+impl VirtioScsiLun {
+    pub fn parse(bytes: [u8; 8]) -> Option<Self> {
+        if bytes == REPORT_LUNS {
+            Some(Self::ReportLuns)
+        } else if bytes[0] == 0x1 {
+            let target = bytes[1];
+            // bytes[2..3] is a normal SCSI single-level lun
+            if (bytes[2] & 0b0100_0000) == 0 {
```

Comment:
Ok, retracing my steps here: The format of the 8 bytes passed in here is defined in the [virtio-scsi spec](https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-308001r311), which says:

> lun addresses the REPORT LUNS well-known logical unit, or a target and logical unit in the virtio-scsi device’s SCSI domain. When used to address the REPORT LUNS logical unit, lun is 0xC1, 0x01 and six zero bytes. The virtio-scsi device SHOULD implement the REPORT LUNS well-known logical unit.
> When used to address a target and logical unit, the only supported format for lun is: first byte set to 1, second byte set to target, third and fourth byte representing a single level LUN structure, followed by four zero bytes. With this representation, a virtio-scsi device can serve up to 256 targets and 16384 LUNs per target. The device MAY also support having a well-known logical units in the third and fourth byte.

The encoding for `bytes[2..3]` appears to match the "Single level LUN structure using flat space addressing method", found in Table 15 of SAM-6 revision 7 (the draft I was developing against) - it's just `01b` followed by a 14-bit LUN number.

Reply:
Oh huh, this code would silently accept `11b` as well - I very much suspect that was a mistake.

Reply:
Ok. Things are making a bit more sense after lunch... I got confused by both structures being 8 bytes, but for virtio it looks like the `0x1, <target>` prefix has little else to do with the LUN in the scsi spec, then the next two bytes are the _actual_ LUN. Will add a check to make sure that we only support `0b01` right now.

Reply:
Incidentally, I have no idea what the "REPORT LUNS well-known LUN" as described in virtio-scsi is meant to do - it doesn't have any place to specify a target, so presumably the intent is that it returns LUNs from all targets? But presumably it would return LUNs in the SCSI format, which doesn't include a target number (because SCSI assumes the command is being sent to a LUN on a single target). In any case, I'm not sure anybody uses it.